// react-native-stack/src/NavigatorIos/NavigationBar/NavigationBarBackgroundIos.js
import React from 'react';
import { Animated, Dimensions, View } from 'react-native';
const ConnectedNavigationBarBackgroundIos = ({ navigationBar, router }) => {
    // const scrollViewContext = scrollViewsContext.getContext(history.location.key)
    let background;
    if (navigationBar.BackgroundComponent &&
        React.isValidElement(navigationBar.BackgroundComponent)) {
        background = navigationBar.BackgroundComponent;
    }
    else if (navigationBar.BackgroundComponent) {
        background = React.createElement(navigationBar.BackgroundComponent, {});
    }
    const screenHeight = Dimensions.get('screen').height;
    return (<Animated.View pointerEvents="box-none" style={{
        height: 0,
        left: 0,
        position: 'absolute',
        right: 0,
        top: 0 /*,
        transform: scrollViewContext.contentInset
          ? [
              {
                translateY: scrollViewContext.scrollAnimatedValue.interpolate({
                  extrapolateRight: 'clamp',
                  inputRange: [
                    screenHeight * 2 * -1,
                    navigationBar.height * -1
                  ],
                  outputRange: [screenHeight * 2, navigationBar.height]
                })
              }
            ]
          : [
              {
                translateY: navigationBar.height
              }
            ]*/
    }}>
      <View pointerEvents="box-none" style={{
        bottom: 0,
        height: Dimensions.get('screen').height * 2,
        left: 0,
        position: 'absolute',
        right: 0
    }}>
        {background}
      </View>
    </Animated.View>);
};
export const NavigationBarBackgroundIos = ConnectedNavigationBarBackgroundIos;
